fs/nfs/nfs4proc.c

```diff
@@ -7771,6 +7775,475 @@ static bool nfs4_xattr_list_nfs4_sacl(struct dentry *dentry)
 	return nfs4_server_supports_acls(NFS_SB(dentry->d_sb), NFS4ACL_SACL);
 }
 
+#if CONFIG_TRUENAS
+/*
+ * We will publish the DACL thru NA41_NAME ("system.nfs4_acl_xdr")
+ *
+ * First some support functions.  These may be similar to upstream
+ * functions in nfsd.
+ */
+
+/*
+ * return the size of the struct nfs4_acl required to represent an acl
+ * with @entries entries.
+ */
+static int nfs4_acl_bytes(int entries)
+{
+	return sizeof(struct nfs4_acl) + entries * sizeof(struct nfs4_ace);
+}
+
+static struct {
+	char *string;
+	int   stringlen;
+	int type;
+} s2t_map[] = {
+	{
+		.string    = "OWNER@",
+		.stringlen = sizeof("OWNER@") - 1,
+		.type      = NFS4_ACL_WHO_OWNER,
+	},
+	{
+		.string    = "GROUP@",
+		.stringlen = sizeof("GROUP@") - 1,
+		.type      = NFS4_ACL_WHO_GROUP,
+	},
+	{
+		.string    = "EVERYONE@",
+		.stringlen = sizeof("EVERYONE@") - 1,
+		.type      = NFS4_ACL_WHO_EVERYONE,
+	},
+};
+
+static int nfs4_acl_get_whotype(char *p, u32 len)
```

Comment:
Since an almost identical function is already defined in `nfsd/nfs4acl.c`, is it possible to reuse that here or define the same data types in some common location?

Reply:
@bmeagherix I agree with you that we should be as little invasive to existing code as possible.

Reply:
@bmeagherix that's fine with me.

Reply:
OK, I will leave this as-is for now.